File: python/KLTWrapper.py
import numpy as np
import cv2
import itertools
feature_params = dict(maxCorners=500,
                      qualityLevel=0.3,
                      minDistance=7,
                      blockSize=7)

# 判断点属于哪个区域
def get_region(point, image_width, image_height):
    x, y = point
    region_width = image_width // 4
    region_height = image_height // 4
    row = y // region_height
    col = x // region_width
    return row * 4 + col
    # return row

# 将属于同一个区域的点放在同一个集合中
def group_points(points0,points1):
    points0_groups = [[] for _ in range(16)]
    points1_groups = [[] for _ in range(16)]
    for i in range(len(points0)):
        region = int(get_region(points0[i], 1920, 1080))
        points0_groups[region].append(points0[i])
        points1_groups[region].append(points1[i])
    return points0_groups,points1_groups

class KLTWrapper:
    def __init__(self):
        self.win_size = 10
        self.status = 0
        self.count = 0
        self.flags = 0

        self.image = None
        self.imgPrevGray = None
        self.H = None

        self.GRID_SIZE_W = 32
        self.GRID_SIZE_H = 24
        self.MAX_COUNT = 0
        self.points0 = None
        self.points1 = None


    def init(self, imgGray):

        (nj, ni) = imgGray.shape

        self.MAX_COUNT = (float(ni) / self.GRID_SIZE_W + 1.0) * (float(nj) / self.GRID_SIZE_H + 1.0)
        self.lk_params = dict(winSize=(self.win_size, self.win_size),
                         maxLevel=3,
                         criteria=(cv2.TERM_CRITERIA_MAX_ITER| cv2.TERM_CRITERIA_EPS, 20, 0.03))
        self. H = np.array([np.identity(3) for _ in range(16)])

    def InitFeatures(self, imgGray):

        self.quality = 0.01
        self.min_distance = 10

        (nj, ni) = imgGray.shape

        self.count = ni / self.GRID_SIZE_W * nj / self.GRID_SIZE_H

        lenI = ni / self.GRID_SIZE_W - 1
        lenJ = nj / self.GRID_SIZE_H - 1
        J = np.arange(lenI*lenJ) / lenJ * self.GRID_SIZE_W + self.GRID_SIZE_W / 2
        I = np.arange(lenJ*lenI) % lenJ * self.GRID_SIZE_H + self.GRID_SIZE_H / 2

        self.points1 = np.expand_dims(np.array(list(zip(J, I))), 1).astype(np.float32)
        self.points0, self.points1 = self.points1, self.points0

    def RunTrack(self, image, imgPrev):

        if self.count > 0:
            self.points1, _st, _err = cv2.calcOpticalFlowPyrLK(imgPrev, image, self.points0, None, **self.lk_params)
            good0 = self.points0[_st == 1]
            good1 = self.points1[_st == 1]
            self.count = len(good0)

        if self.count > 10:
            self.makeHomoGraphy(good0, good1)
        else:
            self.H = np.array([np.identity(3) for _ in range(16)])
        self.InitFeatures(image)

    def makeHomoGraphy(self, p0, p1):
        point0_groups ,point1_groups= group_points(p0,p1)
        a = [np.array([]) for _ in range(16)]
        b = [np.array([]) for _ in range(16)]
        for i in range(16):
          #更改前后向
          b[i]=np.asarray(point0_groups[i])
          a[i]=np.asarray(point1_groups[i])
        if len(a[0])<4:
            H0=np.identity(3)
        else:
            H0, status = cv2.findHomography(a[0], b[0], cv2.RANSAC, 1.0)
            if H0 is None:
               H0=np.identity(3)

        if len(a[1])<4:
            H1 = np.identity(3)
        else:
            H1, status = cv2.findHomography(a[1], b[1], cv2.RANSAC, 1.0)
            if H1 is None:
               H1 = np.identity(3)

        if len(a[2])<4:
            H2 = np.identity(3)
        else:
            H2, status = cv2.findHomography(a[2], b[2], cv2.RANSAC, 1.0)
            if H2 is None:
               H2= np.identity(3)

        if len(a[3])<4:
            H3 = np.identity(3)
        else:
            H3, status = cv2.findHomography(a[3], b[3], cv2.RANSAC, 1.0)
            if H3 is None:
               H3 = np.identity(3)

        if len(a[4])<4:
            H4 = np.identity(3)
        else:
            H4, status = cv2.findHomography(a[4], b[4], cv2.RANSAC, 1.0)
            if H4 is None:
               H4 = np.identity(3)

        if len(a[5])<4:
            H5 = np.identity(3)
        else:
            H5, status = cv2.findHomography(a[5], b[5], cv2.RANSAC, 1.0)
            if H5 is None:
               H5 = np.identity(3)

        if len(a[6])<4:
            H6 = np.identity(3)
        else:
            H6, status = cv2.findHomography(a[6], b[6], cv2.RANSAC, 1.0)
            if H6 is None:
               H6 = np.identity(3)

        if len(a[7])<4:
            H7 = np.identity(3)
        else:
            H7, status = cv2.findHomography(a[7], b[7], cv2.RANSAC, 1.0)
            if H7 is None:
               H7 = np.identity(3)

        if len(a[8])<4:
            H8 = np.identity(3)
        else:
            H8, status = cv2.findHomography(a[8], b[8], cv2.RANSAC, 1.0)
            if H8 is None:
               H8 = np.identity(3)

        if len(a[9])<4:
            H9 = np.identity(3)
        else:
            H9, status = cv2.findHomography(a[9], b[9], cv2.RANSAC, 1.0)
            if H9 is None:
               H9 = np.identity(3)

        if len(a[10])<4:
            H10 = np.identity(3)
        else:
            H10, status = cv2.findHomography(a[10], b[10], cv2.RANSAC, 1.0)
            if H10 is None:
               H10 = np.identity(3)

        if len(a[11])<4:
            H11 = np.identity(3)
        else:
            H11, status = cv2.findHomography(a[11], b[11], cv2.RANSAC, 1.0)
            if H11 is None:
               H11 = np.identity(3)

        if len(a[12])<4:
            H12 = np.identity(3)
        else:
            H12, status = cv2.findHomography(a[12], b[12], cv2.RANSAC, 1.0)
            if H12 is None:
               H12 = np.identity(3)

        if len(a[13])<4:
            H13 = np.identity(3)
        else:
            H13, status = cv2.findHomography(a[13], b[13], cv2.RANSAC, 1.0)
            if H13 is None:
               H13 = np.identity(3)

        if len(a[14])<4:
            H14 = np.identity(3)
        else:
            H14, status = cv2.findHomography(a[14], b[14], cv2.RANSAC, 1.0)
            if H14 is None:
               H14 = np.identity(3)

        if len(a[15])<4:
            H15= np.identity(3)
        else:
            H15, status = cv2.findHomography(a[15], b[15], cv2.RANSAC, 1.0)
            if H15 is None:
               H15 = np.identity(3)
        # H2, status = cv2.findHomography(a[2], b[2], cv2.RANSAC, 1.0)
        # if H2 is None:
        #     H2=np.identity(3)
        # H3, status = cv2.findHomography(a[3], b[3], cv2.RANSAC, 1.0)
        # if H3 is None:
        #     H3=np.identity(3)
        # H4, status = cv2.findHomography(a[4], b[4], cv2.RANSAC, 1.0)
        # if H4 is None:
        #     H4=np.identity(3)
        # H5, status = cv2.findHomography(a[5], b[5], cv2.RANSAC, 1.0)
        # if H5 is None:
        #     H5=np.identity(3)
        # H6, status = cv2.findHomography(a[6], b[6], cv2.RANSAC, 1.0)
        # if H6 is None:
        #     H6=np.identity(3)
        # H7, status = cv2.findHomography(a[7], b[7], cv2.RANSAC, 1.0)
        # if H7 is None:
        #     H7=np.identity(3)
        # H8, status = cv2.findHomography(a[8], b[8], cv2.RANSAC, 1.0)
        # if H8 is None:
        #     H8=np.identity(3)
        # H9, status = cv2.findHomography(a[9], b[9], cv2.RANSAC, 1.0)
        # if H9 is None:
        #     H9=np.identity(3)
        # H10, status = cv2.findHomography(a[10], b[10], cv2.RANSAC, 1.0)
        # if H10 is None:
        #     H10=np.identity(3)
        # H11, status = cv2.findHomography(a[11], b[11], cv2.RANSAC, 1.0)
        # if H11 is None:
        #     H11=np.identity(3)
        # H12, status = cv2.findHomography(a[12], b[12], cv2.RANSAC, 1.0)
        # if H12 is None:
        #     H12=np.identity(3)
        # H13, status = cv2.findHomography(a[13], b[13], cv2.RANSAC, 1.0)
        # if H13 is None:
        #     H13=np.identity(3)
        # H14, status = cv2.findHomography(a[14], b[14], cv2.RANSAC, 1.0)
        # if H14 is None:
        #     H14=np.identity(3)
        # H15, status = cv2.findHomography(a[15], b[15], cv2.RANSAC, 1.0)
        # if H15 is None:
        #     H15=np.identity(3)

        try:
            self.H = np.array([H0, H1, H2, H3,H4,H5,H6,H7,H8,H9,H10,H11,H12,H13,H14,H15])
        except:
            print("error")
        # self.H = np.array([H0, H1, H2, H3])

File: src/KLTWrapper.cpp

#include "KLTWrapper.hpp"
#include <opencv/cv.h>
#include <vector>
#include <opencv2/highgui.hpp>
#include <opencv2/core.hpp>
#include <opencv2/opencv.hpp>
#include <opencv2/imgproc.hpp>
#include "opencv2/calib3d/calib3d.hpp"

using namespace cv;
using namespace std;

KLTWrapper::KLTWrapper(void)
{
}

KLTWrapper::~KLTWrapper(void)
{
}

void KLTWrapper::Init( int image_width, int image_height)
{
	width = image_width;
	height = image_height;
	flags=0;
	for (int j = 0; j < 16; ++j) {
		for (int ii = 0; ii < 9; ++ii) {
			matH[j][ii] = ii % 3 == ii / 3 ? 1.0f : 0.0f;
		}
	}
}

int KLTWrapper::get_region(cv::Point2f point) {
    int x = point.x;
    int y = point.y;
    int region_width = width / 4;
    int region_height = height / 4;
    int row = y / region_height;
    int col = x / region_width;
    return row * 4 + col;
}

vector<vector<vector<Point2f>>> KLTWrapper::group_points(vector<Point2f> good0,vector<Point2f> good1){
    vector<vector<vector<Point2f>>> points_groups(2, vector<vector<Point2f>>(16));
    for(size_t i=0;i<good0.size();i++)
	{
		int region = int(get_region(good1[i]));
		points_groups[0][region].push_back(good0[i]);
        points_groups[1][region].push_back(good1[i]);
    }
return points_groups;
}

void KLTWrapper::InitFeatures()
{
	double quality = 0.01;
	double min_distance = 10;
    prev_pts.clear();
	int ni = width;
	int nj = height;

	count = (int(ni / GRID_SIZE_W)-1) * (int(nj / GRID_SIZE_H)-1);

	int cnt = 0;
	for (int i = 0; i < count; ++i) {
		prev_pts.push_back(cv::Point2f(float(i)/(int(nj / GRID_SIZE_H)-1)* GRID_SIZE_W + GRID_SIZE_W / 2,i%(int(nj / GRID_SIZE_H)-1)* GRID_SIZE_H + GRID_SIZE_H / 2));
	}
}

void KLTWrapper::RunTrack(const UMat& imgGray, const UMat& prevGray)
{
	clock_t star,end,t1,t2,t3,t4,t5,t6;

	int nMatch[MAX_COUNT];
	std::vector<cv::Point2f> good0;
	good0.clear();
	std::vector<cv::Point2f> good1;
	good1.clear();
	if (count > 0) {
		star=clock();
		UMat umatNextPts, umatStatus, umatErr;
		cv::calcOpticalFlowPyrLK(prevGray, imgGray, prev_pts, umatNextPts, umatStatus, umatErr, Size(8,8),1
		,cv::TermCriteria((TermCriteria::COUNT) | (TermCriteria::EPS), 20, (0.03)),0);
		umatNextPts.reshape(2, 1).copyTo(next_pts);
        umatStatus.reshape(1, 1).copyTo(status);
        umatErr.reshape(1, 1).copyTo(err);
        end=clock();

		t1=clock();
		for (int i = 0; i < prev_pts.size(); i++) {
			if (!status[i]||next_pts[i].x<0||next_pts[i].y<0||next_pts[i].x > width || next_pts[i].y>height) {
				continue;
			}
			good0.push_back(prev_pts[i]);
            good1.push_back(next_pts[i]);
		}
		t2=clock();
	}
    t3=clock();
	if (count >= 10) {

		MakeHomoGraphy(good0, good1);
	} else {
		for (int j = 0; j < 16; ++j) {
		for (int ii = 0; ii < 9; ++ii) {
			matH[j][ii] = ii % 3 == ii / 3 ? 1.0f : 0.0f;
		}
		}
	}
	t4=clock();

	t5=clock();
    InitFeatures();
	t6=clock();
// double ti=end-star;	
// double t=t2-t1;
// double tt=t4-t3;
// double ttt=t6-t5;
// std::cout<<"calcop:"<<ti/CLOCKS_PER_SEC<<" "<<"sec_point:"<<t/CLOCKS_PER_SEC<<" "<<"MakeHomoGraphy:"<<tt/CLOCKS_PER_SEC<<" "<<"InitFeatures:"<<ttt/CLOCKS_PER_SEC<<std::endl;
}

void KLTWrapper::GetHomography(double (*h)[9])
{
	memcpy(h, matH, sizeof(matH));
}

void KLTWrapper::MakeHomoGraphy(std::vector<cv::Point2f>&good0,std::vector<cv::Point2f> &good1)
{
	vector<vector<vector<Point2f>>> point_groups=group_points(good1,good0);
	double h[16][9];
	vector <Mat> _h(16, Mat(3, 3, CV_64F));

for(int j=0;j<16;j++){
	if(point_groups[0][j].size()<4)
	{
		for (int ii = 0; ii < 9; ++ii) {
			matH[j][ii]= ii % 3 == ii / 3 ? 1.0f : 0.0f;
		}
	}else{
		_h[j]=findHomography(point_groups[0][j], point_groups[1][j], RANSAC, 1);
		if(_h[j].empty()){
           for (int ii = 0; ii < 9; ++ii) {
			   matH[j][ii]= ii % 3 == ii / 3 ? 1.0f : 0.0f;
		    }
		}else{
		   for (int ii = 0; ii < 9; ++ii) {
			matH[j][ii]= _h[j].at<double>(ii);
		}
		}
	}
	}
}
